### PR TITLE
Fix bugs in merge command

### PR DESCRIPTION
**Pull request by sonalideshpandemsft**
The sync-branch workflow in the pipeline at https://github.com/microsoft/FluidFramework/actions/workflows/sync-branch.yml is encountering failures. I was able to replicate the error locally, and here are the errors along with their corresponding solutions.

1. Error: "branch name required" - To fix this, update the reference to refs/remotes/origin/next instead of just "next" when passing it as a branch name.

2. Error: "Committer identity unknown" - To resolve this, configure Git to check the merge operation by setting the global user email and name using the commands:
`git config --global user.email "you@example.com"`
`git config --global user.name "Your Name"`

3. Error: "could not read Username for 'https://github.com/': No such device or address" - This error occurred while pushing the branch to the remote repository. To fix it, set the Git configuration using the following command, which substitutes the default GitHub URL with a custom URL that includes a secret token:
`await this.gitRepo.gitClient.raw([
"config",
"--global",
`url.https://${secret}@github.com/.insteadOf`,
"https://github.com/",]);`

In summary, the `url.https://${secret}@github.com/.insteadOf` configuration is used to substitute the default GitHub URL with a custom URL that includes a secret token for secure authentication and authorization during Git operations, such as pushing to a remote repository on GitHub.

4. The GitHub workflow fails when there is an existing open PR. Fixed the workflow to not error out.

[AB#4873](https://dev.azure.com/fluidframework/internal/_sprints/taskboard/Team%20Sasha/internal/Team%20Sasha%20-%202023.06-B?workitem=4873)

### CODE REVIEW

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 1]
```diff
@@ -55,6 +48,11 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 			required: true,
 			multiple: true,
 		}),
+		creds: Flags.string({
+			description: "Username and email id for GitHub auth",
+			required: true,
+			multiple: true,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:54]: This should be false, no?

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:54]: Oh I see... I suggest separate flags. That's more obvious, and otherwise order matters, which is confusing. You can parse the two flags into a single object in the run method and then pass that around as creds. E.g.

```ts
interface GitHubCredentials {
  username: string, 
  email: string,
}

...

const creds: GitHubCredentials = { username: flags.username, email: flags.email };

```

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 2]
```diff
@@ -182,11 +186,19 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
 			`Creating and checking out branch: ${branchName} at commit ${shortCommit(
 				prHeadCommit,
 			)}`,
 		);
+
+		await this.gitRepo.gitClient.raw([
+			"config",
+			"--global",
+			`url.https://${flags.pat}@github.com/.insteadOf`,
+			"https://github.com/",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:199]: This is not obvious, so I'd put the info from the PR description in a comment. Is there documentation on this behavior somewhere that we could link to as well?

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 3]
```diff
@@ -107,14 +104,17 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 			// TODO: notify the author
 		}
 
-		const lastMergedCommit = await this.gitRepo.getMergeBase(flags.source, flags.target);
+		const lastMergedCommit = await this.gitRepo.getMergeBase(
+			`${flags.source}`,
+			`refs/remotes/${this.remote}/${flags.target}`,
+		);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:110]: Use getMergeBaseRemote instead.

File: build-tools/packages/build-cli/src/lib/git.ts [hunk 4]
```diff
@@ -193,14 +193,17 @@ export class Repository {
 			.filter((value) => value !== null && value !== undefined && value !== "");
 	}
 
-	public async canMergeWithoutConflicts(commit: string): Promise<boolean> {
+	public async canMergeWithoutConflicts(commit: string, creds: string[]): Promise<boolean> {
 		let mergeResult;
+		await this.git.addConfig("user.email", creds[0]);
+		await this.git.addConfig("user.name", creds[1]);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/git.ts:199]: I would not do this here. This function shouldn't need to configure the repo. Instead, this function should assume that these config settings have been set. Then you can just set them once in the run method and not pass the creds around - you just use the flags to set the repo settings once.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 5]
```diff
@@ -138,11 +138,15 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 		this.branchesToCleanup.push(tempBranchToCheckConflicts);
 
 		await this.gitRepo.gitClient
-			.checkoutBranch(tempBranchToCheckConflicts, flags.target)
-			.branch(["--set-upstream-to", `${this.remote}/${tempBranchToCheckConflicts}`]);
+			.checkoutBranch(
+				tempBranchToCheckConflicts,
+				`refs/remotes/${this.remote}/${flags.target}`,
+			)
+			.branch(["--set-upstream-to", `${tempBranchToCheckConflicts}`]);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:145]: This doesn't seem right. This is creating a branch foo-automation at refs/remotes/${this.remote}/${flags.target}, but then setting the upstream of that branch to foo-automation. Maybe you meant this?

```suggestion
			.checkoutBranch(
				tempBranchToCheckConflicts,
				`refs/remotes/${this.remote}/${flags.target}`,
			)
			.branch(["--set-upstream-to", `${this.remote}/${tempBranchToCheckConflicts}`]);
```

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:145]: But why does an upstream branch have to be set at all? This is just a temporary branch used locally, right?

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 6]
```diff
@@ -55,6 +48,11 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 			required: true,
 			multiple: true,
 		}),
+		creds: Flags.string({
+			description: "Username and email id for GitHub auth",
+			required: true,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:53]: You don't want to require these if you're going to set the git config settings in the command code. If you do that, then when testing locally, people will have to 

I think you're basically in a position where the repo has to be "configured properly" in order to use the command. That's totally fine. The question is where should that config happen? I think the best place is in the workflow itself.

Imagine two scenarios - running the command locally for testing and running in CI. When running locally, you would assume safely that the user has their repo configured with a username and email. When in CI, it's sort of like the bot running the command locally - so it's reasonable to expect "the bot" to configure their repo as well. So I think setting the username/email in the workflow before calling the flub command makes more sense and will be simpler code overall.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts (hunk 5 above)
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:145]: yeah. I should not set temporary branch to upstream.

File: build-tools/packages/build-cli/src/lib/git.ts (hunk 4 above)
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/lib/git.ts:199]: Setting them in the GitHub workflow

File: build-tools/packages/build-cli/src/commands/merge/branches.ts (hunk 2 above)
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:199]: I don't have link to the documentation, I found a solution similar to your command through Stack Overflow. This can be set in the workflow as well.

File: build-tools/packages/build-cli/src/lib/git.ts [hunk 7]
```diff
@@ -103,16 +103,6 @@ export class Repository {
 		return base;
 	}
 
-	/**
-	 * @param ref1 - The first ref to compare.
-	 * @param ref2 - The ref to compare against.
-	 * @returns The ref of the merge base between the two refs.
-	 */
-	public async getMergeBase(ref1: string, ref2: string): Promise<string> {
```
Comment (sonalideshpandemsft) [inline build-tools/packages/build-cli/src/lib/git.ts:111]: not used anywhere

File: .github/workflows/sync-branch.yml [hunk 8]
```diff
@@ -33,6 +37,15 @@ jobs:
           # add the bin dir of build-cli to the path
           cd packages/build-cli/bin
           echo "$(pwd)" >> $GITHUB_PATH
+
+      # the `url.https://${secret}@github.com/.insteadOf` configuration is used to substitute the default GitHub URL 
+      # with a custom URL that includes a secret token for secure authentication and authorization during Git operations, 
+      # such as pushing to a remote repository on GitHub.
```
Comment (tylerbutler) [inline .github/workflows/sync-branch.yml:43]: The docs for this setting are here: https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf

And this was helpful for me to understand how these settings look in the git config file: https://gist.github.com/Kovrinic/ea5e7123ab5c97d451804ea222ecd78a

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 9]
```diff
@@ -107,14 +99,18 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 			// TODO: notify the author
 		}
 
-		const lastMergedCommit = await this.gitRepo.getMergeBase(flags.source, flags.target);
+		const lastMergedCommit = await this.gitRepo.getMergeBaseRemote(
+			flags.source,
+			this.remote,
+			`refs/remotes/${this.remote}/${flags.target}`,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:105]: I find the logic a lot easier to follow if you use local branches wherever possible. You don't have to change this necessarily, but if you're going to use strings like refs/remotes/${this.remote}/${flags.target}, then I suggest making them variables with names like remoteTargetBranch. That makes it more obvious what the string represents.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 10]
```diff
@@ -219,7 +217,7 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 
 		if (prWillConflict === false) {
 			await this.gitRepo.gitClient
-				.merge([flags.target, "-m", prTitle])
+				.merge([`${this.remote}/${flags.target}`, "-m", prTitle])
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:220]: This is confusing to me again. Why do you want to merge with the remote branch instead of the local branch?

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:220]: If you're trying to ensure you merge with the latest, you would have to do a `git fetch` to update the remote refs anyway, so merging with the remote ref doesn't really save you anything.

File: build-tools/packages/build-cli/src/lib/git.ts [hunk 11]
```diff
@@ -198,9 +198,9 @@ export class Repository {
 		try {
 			console.log(`Checking merge conflicts for: ${commit}`);
 			mergeResult = await this.git.merge([commit, "--no-commit", "--no-ff"]);
-			console.log(`Git merge result: ${mergeResult}`);
 			await this.git.merge(["--abort"]);
 		} catch {
+			console.log(`Merge result conflicts exists for: ${commit}`);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/git.ts:203]: nit:

```suggestion
			console.log(`Merge conflicts exists for: ${commit}`);
```

File: build-tools/packages/build-cli/docs/merge.md [hunk 12]
```diff
@@ -12,12 +12,9 @@ Sync branches depending on the batch size passed
 
 ```
 USAGE
-  $ flub merge branches -a <value> -p <value> -s <value> -t <value> -b <value> --reviewers <value> [-v | --quiet]
-    [-r <value>]
+  $ flub merge branches -p <value> -s <value> -t <value> -b <value> --reviewers <value> [-v | --quiet] [-r <value>]
 
 FLAGS
-  -a, --auth=<value>       (required) GitHub authentication token. For security reasons, this value should be passed
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/docs/merge.md:19]: To confirm, you still have to provide a PAT, right? Can the PAT be provided using an env variable? If not, we should add that. For security reasons, you generally don't want to pass a token via the CLI, because terminals have history, and your PAT winds up in the terminal logs, which can disclose your token unexpectedly. With an env variable, what people usually do is put the token in a file, then read that file into an env variable, and read the token from the variable.

To be clear, this is only relevant for local testing. The bot/workflow is fine. But supporting an env variable will make local testing much more secure.

**Review by tylerbutler — APPROVED** [review]
I had some more feedback, but I'm approving so you can use your judgement about what feedback to take before merge. Most of my feedback is minor, so I would be OK with doing some of them in a followup. Your call.

File: build-tools/packages/build-cli/docs/merge.md (hunk 12 above)
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/docs/merge.md:19]: I did setup the environment variables using the env keyword in GitHub Actions which will not expose the values in the terminal logs by default.
